### ISSUE TITLE
Excess CPU usage

### ISSUE THREAD

**Issue by Symbitic** [pulsar-edit/pulsar#992]
### Thanks in advance for your bug report!

- [X] Have you reproduced issue in [safe mode](https://pulsar-edit.dev/docs/atom-archive/faq/sections/what-does-safe-mode-do.html)?
- [X] Have you used the [debugging guide](https://pulsar-edit.dev/docs/atom-archive/hacking-atom/#debugging) to try to resolve the issue?
- [X] Have you checked our [FAQs](https://pulsar-edit.dev/docs/launch-manual/sections/faq/) to make sure your question isn't answered there?
- [X] Have you checked to make sure your issue does not already [exist](https://github.com/issues?q=is%3Aissue+user%3Apulsar-edit+is%3Aopen)?
- [X] Have you checked you are on the latest release of Pulsar?

### What happened?

As soon as I open pulsar, the heat skyrockets and my fans go to 100%. `top` says pulsar is using 99.7% of the CPU. As soon as I close pulsar, it only takes a few seconds for CPU usage to die down and fans to fall silent.

### Pulsar version

1.116.0

### Which OS does this happen on?

🐧 Red Hat based (Fedora, Alma, RockyLinux, CentOS Stream, etc.)

### OS details

Fedora 40 KDE

### Which CPU architecture are you running this on?

x86_64/AMD64

### What steps are needed to reproduce this?

1. Open pulsar
2. Wait a few seconds.

### Additional Information:

This seems fairly recent as I have been using atom/pulsar for many years without this problem. I'm running Fedora KDE 40 on my Samsung Galaxy Book 360.

**Comment by Xon77** [comment]
Hello,

I've noticed also excess CPU usage on Apple Silicon ARM 1 with OS 14.7.1, not only at startup, but also when the editor is running.

I often use either Tidal alone or Hydra, and the editor freezes a few seconds relatively often (every 5 minutes).

When I look at the Activity Monitor, Pulsar Helper (Renderer) sometimes goes above 100% for a few seconds, which makes it freeze. 

It makes Pulsar almost unusable with Hydra at least on Mac Sillicon.

Does anyone know how to solve this ?

Thanks

**Comment by Xon77** [comment]
You can find here the profile data where there is a task at the end which freezes the editor.
Hope it helps.
[Profile-20241210T193551.json.zip](https://github.com/user-attachments/files/18087824/Profile-20241210T193551.json.zip)

**Comment by savetheclocktower** [comment]
Can you test this with either Tidal or Hydra (or both) disabled? Would be good to know whether the culprit is one of those packages or the editor itself.

**Comment by Xon77** [comment]
I can experience freeze and excess CPU usage, with Tidal alone or Hydra alone, and even with Pulsar alone (less often). It does not seem to be packages. Does the profile help?

**Comment by savetheclocktower** [comment]
I haven't had time to look at the profile yet, but I'll see if I can do so this week.

**Comment by Xon77** [comment]
Thank you so much for your help!

**Comment by Xon77** [comment]
@savetheclocktower and Others.... Did you have some time to identify the reason of CPU usage on Mac ? This excess CPU makes the editor unusable in live performances contexts to me. I can test more and provide more tests if you tell me what to do. Sorry I am not a professional developer.

**Comment by savetheclocktower** [comment]
Sorry for the delay! The big hiccup in that profile data appears to be related to refreshing Git status on a project.

If you have time to test it, you might want to attempt to work an a project of comparable size to the one that gives you problems, but without a `.git` folder so that Pulsar does not treat it as a Git project. (You could just copy the folder to a temporary location and delete `.git` from the temporary version.) Make sure no other projects are open in Pulsar.

There's another issue that affects Pulsar on Silicon Macs — I know because it affected me: any task that requires shelling out via Node (running a process in the background) will freeze for a fraction of a second. This is a bug in the version of `libuv` used by Electron version 12, and the only fix for this issue is to use a newer version of Electron.

We're working on that; we've got a new version of Pulsar under active development, and you'd be able to [try it here](https://github.com/pulsar-edit/pulsar-electron-next-binaries/releases/tag/v1.124.2025010601-next) — except that we don't routinely build Apple Silicon binaries for this pre-release version for complicated reasons. You might still try the Intel version (which can run under Rosetta) and see if it's to your liking, and whether you can install `hydra` and/or `tidalcycles` and have them work properly. I could also walk you through building a version for Apple Silicon from source if you had lots of time on your hands and were particularly eager.

**Comment by Symbitic** [comment]
I was never using Pulsar as a project editor; only a text editor. Mostly I had files open in my Downloads folder, but I may have had some files open that were nested inside a folder with a `.git`. Could that have been leading it to try refreshing the git status of every subdirectory it could find?

**Comment by savetheclocktower** [comment]
> I was never using Pulsar as a project editor; only a text editor. Mostly I had files open in my Downloads folder, but I may have had some files open that were nested inside a folder with a `.git`. Could that have been leading it to try refreshing the git status of every subdirectory it could find?

Possibly, but doesn't seem likely. If you feel like taking 20 minutes to run a quick test, you could find out pretty quickly.

If you can still reproduce the CPU usage, you might also want to take a profile via the Chrome dev tools (<kbd>Ctrl+Shift+I</kbd>, choose Performance, hit the circle/record button in the toolbar, do some stuff that triggers CPU intensive work, stop the profile and save it somewhere). I regret not giving you this advice back when you opened this ticket in April.

**Comment by Symbitic** [comment]
There's nothing that triggers the CPU usage. It happens as soon as Pulsar starts, even if no files are open.

**Comment by savetheclocktower** [comment]
> There's nothing that triggers the CPU usage. It happens as soon as Pulsar starts, even if no files are open.

In that case, you should be able to run `pulsar --profile-startup` to create a startup profile. The developer tools should start in the open state; once you've got a good sample, you can stop the profile and save it somewhere, then attach it to this issue.

**Comment by Xon77** [comment]
@savetheclocktower Thanks a lot for your feedback!

I do not understand why "the big hiccup in the profile data appears to be related to refreshing Git status on a project", since I do not use git in the files I used in Pulsar, only files connected to play with Tidal or Hydra.
Maybe, I should send you another profile data?

I've tried the new Pulsar version, although it is compiled for Intel. Unfortunately, I have the same problem of excess CPU usage, which makes me doubt that the next Apple Silicon binary will solve this problem of freezing 1 or 2 minutes each 3 or X minutes roughly. In the activity monitor, I still see PulsarNext Helper (Renderer) going above 100% each X minutes, and the freeze happens.

Maybe I could send you a profile data with this new Intel version, if it may help?
And should I open a new issue, since at least for me on Apple Silicon, I have this issue relatively often, not only at startup?

**Comment by savetheclocktower** [comment]
> Maybe I could send you a profile data with this new Intel version, if it may help? And should I open a new issue, since at least for me on Apple Silicon, I have this issue relatively often, not only at startup?

Yes in both instances. Thanks for being so willing to help us troubleshoot!

**Comment by Xon77** [comment]
[Trace-20250114T170758.json.zip](https://github.com/user-attachments/files/18416704/Trace-20250114T170758.json.zip)

Here you can find the profile data, where there is excess CPU usage at the end with the new version of Pulsar Intel. Hope it helps. I will open a new issue tonight. Thanks

**Comment by savetheclocktower** [comment]
After looking at this stack trace, I'm not entirely sure what to say. It largely looks great, except every few seconds there's a small hiccup.

The whole thing reflects a lot of steady work being done — which makes sense if it's drawing a visualization. I can say that it appears to be doing a _lot_ of work for each animation frame; there are four calls to `draw` for each “task,” meaning each bundle of work being done after being scheduled via `requestAnimationFrame`. Because a screen ideally draws 60 times in 1000 milliseconds, you want a task to take no more than one frame of work, or 16.66667ms.

Most of these tasks fit in the allotted time, but a few take several times longer than that.

I want to be able to say that Pulsar is adding some sort of overhead here, but I don't see any obvious culprits. There are a couple of places in this profile where the pauses are extremely long, and in those instances garbage collection is to blame.

But the vast majority of work is being done in `typedArg.value`, [which you can see here](https://github.com/hydra-synth/hydra-synth/blob/3163bbd01b597d0bd51babec2906d28331b6a62c/src/format-arguments.js#L66-L79). This appears to be a function that inspects user input, figures out which parameters are functions, and then runs those functions to determine which values to use.

My best theory here is that those functions are throwing errors. Exceptions are costly, and it would explain why I see some code relating to formatting of a stack trace. That code doesn't take long, but the fact that it has to be run at all suggests that you're hitting the `catch` block of this function.

I would check your console to see if you have lots of warnings called `ERROR` listed; if so, that would be one explanation for how this function could take so long. You'd expect those to be transient — something that happens a few times while a user is typing — but if I'm right about this, the `catch` block is getting reached _every time_, even when the animation seems to be running smoothly.

Most of the CPU time really does appear to be consumed by Hydra's own work — either drawing to the screen or doing something else that's not really visible. And it is CPU time — the GPU is barely doing any work.

There is some Pulsar code that runs to try to reformat stack traces on errors, but the profile doesn't identify it as a major contributor. It takes a very brief amount of time whenever it runs — about 1/8 of a millisecond.

---

Some other things you can try:

1. Seriously, leave your dev tools console open as you code in `atom-hydra` and see if it's generating lots of messages. That alone could explain the slowness.
2. Compare performance between `atom-hydra` and the same task performed in [web hydra](https://hydra.ojack.xyz/?sketch_id=example_15) — if `atom-hydra` is way slower, then that opens up a new mystery.
3. Do another profile, but this time check the “memory” checkbox so we can get a sense of whether memory usage is a possible culprit.
4. If you can consistently reproduce the freeze without Tidal or Hydra enabled (tall task), you could capture another profile and send it along.

Meanwhile, I'll do what I can to get you an Apple Silicon build of PulsarNext; it's the next step in this process and it's something we should be doing anyway, since GitHub Actions now has Silicon runners.

**Comment by Xon77** [comment]
@savetheclocktower Thanks a lot for your answer. Sorry I could not test completely before.

I have no issue of freezing with Web Hydra.

But, by looking more carefully at the console, I've discovered some error messages below that are triggered exactly at the moment when it freezes. 

I use Tidal with SuperCollider and Hydra receives RMS data from SuperCollide.

Maybe, do you have an idea about these messages, their reasons and how to correct them ?
It would be wonderful if I could solve this freeze issue.

error SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at osc-loader.js:77
    at Array.forEach (<anonymous>)
    at Socket.OscLoader.stdout (osc-loader.js:74)
    at Socket.emit (events.js:315)
    at addChunk (internal/streams/readable.js:309)
    at readableAddChunk (internal/streams/readable.js:284)
    at Socket.Readable.push (internal/streams/readable.js:223)
    at Pipe.onStreamRead (internal/stream_base_commons.js:188)
osc-loader.js? [sm]:78 error SyntaxError: Unexpected token , in JSON at position 17
    at JSON.parse (<anonymous>)
    at osc-loader.js:77
    at Array.forEach (<anonymous>)
    at Socket.OscLoader.stdout (osc-loader.js:74)
    at Socket.emit (events.js:315)
    at addChunk (internal/streams/readable.js:309)
    at readableAddChunk (internal/streams/readable.js:284)
    at Socket.Readable.push (internal/streams/readable.js:223)
    at Pipe.onStreamRead (internal/stream_base_commons.js:188)

The JSON at position X may change with other freezes. But each time, it freezes, I have this.

Thanks for your help. And I hope someone could help me.

**Comment by savetheclocktower** [comment]
@Xon77, I don't have any immediate insights, but if I were troubleshooting this myself I'd use the developer tools.

Open the dev tools again and switch to the Source tab. Press <kbd>Cmd + O</kbd> (with focus in the dev tools) to open a file picker; type `osc-loader` and select whatever option is offered to you. Go to line 77. That's where the failure is happening. It should [look like this](https://github.com/hydra-synth/atom-hydra/blob/master/lib/osc-loader.js#L77).

You should be able to enable the “Pause on exceptions” option (the octagon with the “pause“ symbol on it) and, if you're lucky, you might then find that the debugger has stopped on that line.

The `atom-hydra` package is using a client/server model. The server is running in a separate node process; the client is running within Pulsar. They're sending messages to one another by watching each other's output. The client seems to be assuming that it's [NDJSON](https://en.wikipedia.org/wiki/JSON_streaming#NDJSON) — a series of discrete JSON messages with newlines in between. That's why it's splitting the received output on `\n` and assuming that each chunk will be valid JSON.

It sounds like that assumption is routinely being violated. If you can get the debugger to stop on such an error, you can find out what the input is. Hover over the `s` variable, or type `s` in the console while the debugger is paused, in order to find out what the message is.

Two theories:

1. The client seems to think that, whenever it receives input, it will always be fully-formed JSON. This may or may not be true. It's possible that the `stdout` handler gets called in the middle of `osc-server.js` sending an especially long message; I forget what the default behavior is in Node in this situation.
2. If the server ever routinely sent a message to its output that was _not_ valid JSON, then the client would routinely be attempting and failing to parse that message. This would explain the frequency with which these errors are generated, but I don't see any obvious way that that could happen based on how the server is set up.

Between these two theories, theory 1 is more plausible to me. If you are able to prove that it's trying to act on a fragment of a message, then someone could theoretically fork `atom-hydra` and fix it without too much effort. I was able to use [ndjson](https://www.npmjs.com/package/ndjson) in another Pulsar package to handle gotchas like this.

But it's just as likely to be something I haven't thought of!

**Comment by Xon77** [comment]
@savetheclocktower Thanks for your input.

Yes, the debugger has stopped on line 77.
Tou can find below the screenshot showing the input that triggered the error.
But I cannot see any thing wrong with the format myself. 

Do you what I could do next to solve this issue  or how to ignore fully-formed JSON ?

<img width="1618" alt="Image" src="https://github.com/user-attachments/assets/3baec1cd-5295-43fd-b224-5db6f14b411b" />

**Comment by savetheclocktower** [comment]
If that's the entire line, then that's your problem — it's looking at a JSON structure that has started but has not yet finished. This is probably because it's trying to process before all of the message is sent, rather than because there's an arbitrary newline in the middle of the message.

To find out for certain, switch to the console tab while your debugger is paused and type `s` in the console to see the full value before splitting. If it's equal to `line`, then that's the problem — processing too early. If I'm wrong, and there's an actual newline in the middle of the sent JSON, then `osc-server.js` needs to be fixed so that it does not emit JSON with a newline in the middle of it.

But let's assume it's the first thing for now.

I theoretically can advise on how to fix it, but you might not have enough experience to do it, and I might not have enough time. Still, let's give it a shot. The fix is to implement more robust handling of JSON message-passing with a library like [ndjson](https://www.npmjs.com/package/ndjson).

First: make a backup of the entire `atom-tidal` package folder so you can revert a file if something blows up.

Second: Run `npm install ndjson` from the root of the `~/.pulsar/packages/atom-hydra` folder in your terminal. (If you don't have `npm`, install Node; any version of Node should give you `npm` in the terminal.)

Next, open `osc-loader.js` for editing. Near the top of the file you'll want to add

```js
import ndjson from 'ndjson';
```

Line 44 ([use this](https://github.com/hydra-synth/atom-hydra/blob/master/lib/osc-loader.js) as your line number reference, since your local line numbers will be different now) should look like this instead of what's currently there:

```js
this.server.stdout.pipe(ndjson.parse({ strict: false })).on('data', this.stdout);
```

And all of the `stdout` function — from lines 68 to 101 — can be replaced with this much simpler version:

```js
stdout = (msg) => {
  this.emit('*', msg)
  this.emit(msg.address, msg.args)
};
```

Once you've made these changes, save them and reload your Pulsar window. If a miracle has happened, and I have correctly diagnosed this from afar, then you might enjoy a much more performant `atom-hydra` package, and we could discuss formalizing this change via forking the package and publishing it to PPM.

**Comment by Xon77** [comment]
@savetheclocktower I've tried to make all these changes and follow them.
But when I try to toggle Hydra again, Ive got this error below.
Do you have an idea what I do wrong ? Thanks for all your effort!

Failed to activate the atom-hydra package

Cannot use import statement outside a module
[Hide Stack Trace](https://github.com/pulsar-edit/pulsar/issues/992#)
/Users/xon/atom-hydra/node_modules/hydra-synth/src/index.js:1
(function (exports, require, module, __filename, __dirname, process, global, Buffer) { return function (exports, require, module, __filename, __dirname) { import Synth from './hydra-synth.js'
                                                                                                                                                           ^^^^^^
SyntaxError: Cannot use import statement outside a module
    at new Script (vm.js:101:7)
    at NativeCompileCache.runInThisContext (/Applications/Pulsar.app/Contents/Resources/app.asar/src/native-compile-cache.js:41:20)
    at Module._compile (/Applications/Pulsar.app/Contents/Resources/app.asar/src/native-compile-cache.js:99:36)
    at Object.value [as .js] (/Applications/Pulsar.app/Contents/Resources/app.asar/src/compile-cache.js:252:23)
    at Module.load (internal/modules/cjs/loader.js:935:32)
    at Module._load (internal/modules/cjs/loader.js:776:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12913)
    at Function.o._load (electron/js2c/renderer_init.js:33:379)
    at Module.require (internal/modules/cjs/loader.js:959:19)
    at require (/Applications/Pulsar.app/Contents/Resources/app.asar/src/native-compile-cache.js:67:27)
    at Object.<anonymous> (main.js:12:1)
    at Object.<anonymous> (main.js:281:2)
    at Module._compile (/Applications/Pulsar.app/Contents/Resources/app.asar/src/native-compile-cache.js:120:30)
    at Object.value [as .js] (/Applications/Pulsar.app/Contents/Resources/app.asar/src/compile-cache.js:252:23)
    at Module.load (internal/modules/cjs/loader.js:935:32)
    at Module._load (internal/modules/cjs/loader.js:776:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12913)
    at Function.o._load (electron/js2c/renderer_init.js:33:379)
    at Module.require (internal/modules/cjs/loader.js:959:19)
    at require (/Applications/Pulsar.app/Contents/Resources/app.asar/src/native-compile-cache.js:67:27)
    at Object.<anonymous> (atom-hydra.js:3:1)
    at Object.<anonymous> (atom-hydra.js:49:2)
    at Module._compile (/Applications/Pulsar.app/Contents/Resources/app.asar/src/native-compile-cache.js:120:30)
    at Object.value [as .js] (/Applications/Pulsar.app/Contents/Resources/app.asar/src/compile-cache.js:252:23)
    at Module.load (internal/modules/cjs/loader.js:935:32)
    at Module._load (internal/modules/cjs/loader.js:776:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12913)
    at Function.o._load (electron/js2c/renderer_init.js:33:379)
    at Module.require (internal/modules/cjs/loader.js:959:19)
    at require (/Applications/Pulsar.app/Contents/Resources/app.asar/src/native-compile-cache.js:67:27)

**Comment by savetheclocktower** [comment]
Looks like `atom-hydra` isn't pinning to an exact version of `hydra-synth`, and `hydra-synth` upgraded from CommonJS to ES Modules in a patch version (a semver no-no).

* Go to `~/.pulsar/packages/atom-hydra`
* Edit `package.json` and change `"hydra-synth": "^1.3.18",` to `"hydra-synth": "1.3.18",` (remove the `^`)
* Run `ppm install` from within the `atom-hydra` directory (hopefully that's all you need to do to trigger the reinstallation of `hydra-synth`
* Reload your Pulsar window

Unrelated problem, but I bet it tried to install the latest `hydra-synth` when you added `ndjson`.

**Comment by Xon77** [comment]
@savetheclocktower Thanks for your help !

Well, now I can toggle Hydra and have no longer error messages, when it freezes.
Unfortunately, as said in the previous sentence, the freeze is always there.

I tried to record the profile data again with the memory I attached.
Sorry to disturb you with this issue. I would really would like to fix it.
I hope the profile data will help to understand this freeze.

Is there anything I can do to help to define it more precisely ?

Maybe, it is also linked to Tidal ?
But, I have also experienced freeze with Hydra alone (even if it seems less often), and even Tidal alone (if my memory does not deceive me) !
Very difficult for me to test it without Tidal and/or Hydra because the only reason I use Pulsar is because of these both extensions.

Thanks anyway !

[Profile-20250223T121008.json.zip](https://github.com/user-attachments/files/18932042/Profile-20250223T121008.json.zip)

**Comment by savetheclocktower** [comment]
@Xon77 The profile you sent me shows what looks like a very smooth session interrupted by a freeze of ~3 seconds. Is that what you experienced?

If so, then I'm a bit baffled. The profiler says 3.34 seconds were spent on “Function Call,” but it does not identify where the function is or even its name.

About a third of that time was spent dealing with side effects of that function (whatever it is).

Try this:

* Open developer tools and focus the Sources tab.
* Hit <kbd>Cmd + O</kbd> and type in `git-repository`; choose the first option.
* Scroll down to a method called `refreshStatus` — line 558 for me, but your line number may vary.
* Find the line that reads `const relativeProjectPaths =` and click on its line number in the gutter. The breakpoint might get placed on the following line; that's fine.
* Right-click on the breakpoint and select Edit Breakpoint…
* Type this in exactly: `console.log('repo:', repo.getPath()); false` and hit <kbd>Return</kbd>.
* Reproduce whatever causes the freeze; be sure to keep the developer tools open.

Theoretically, this will at least reveal whatever Git repository is having its status refreshed; this is among the aforementioned side-effects of the mystery function. You say the work you're doing is not inside a Git repo, so this feels like a strange bug. `refreshStatus` ends up doing deep equality checks of some objects; that alone takes up ~0.75s, so those objects must be large.

I'm hoping this will shed some light on just what is happening during that freeze.

**Comment by Xon77** [comment]
@savetheclocktower Thanks for the perseverance.
Below the screenshot! I hope I followed your instructions correctly?

<img width="1718" alt="Image" src="https://github.com/user-attachments/assets/ba2cb4ba-c456-42ff-a14e-d0376c56015a" />

The work I am doing is not inside a Git repo. It is just Tidal and Hydra code.
However, inside the packages of tidalCycles and Hydra, there are some git repositories. 
Just to be clear, I made screenshots of my folders of tidalCycles and Hydra below.

<img width="1310" alt="Image" src="https://github.com/user-attachments/assets/a3797abe-05ab-45ef-a741-8b259ec2cd64" />

<img width="1308" alt="Image" src="https://github.com/user-attachments/assets/9fb2e955-6a32-4065-a7ea-3a8ca4e092e1" />

I even tried to cut my wifi in case desperately to check if the freeze was also present.
Unfortunately it still is present whatever I do...

**Comment by savetheclocktower** [comment]
That's not quite it. A regular breakpoint has a blue gutter marker, but a conditional breakpoint will have an orange marker.

The goal is to get it to log which Git repo it's refreshing the status of in the console without actually stopping the code.

So try again and then send a screenshot of the Console tab's output. Make sure the marker is orange and that the code doesn't actually pause, but rather puts logging statements into the Console tab. You should see some logging lines that say `repo:` followed by a path on disk.

**Comment by savetheclocktower** [comment]
I was imprecise in a couple of places, so I updated the directions in my earlier comment.

**Comment by Xon77** [comment]
Sorry, below the new screenshot with the repo showed : 

repo: /Users/xon/.git/
repo: /Users/xon/atom-hydra/.git/

<img width="1218" alt="Image" src="https://github.com/user-attachments/assets/ef93607e-2239-4c86-b89f-e75a4ab64dd3" />

**Comment by savetheclocktower** [comment]
If `atom-hydra` truly lives inside of `~/.pulsar/packages/atom-hydra` and doesn't just symlink to `~/atom-hydra`, then I'm not sure why the latter's Git repository status is being refreshed.

The fact that your entire home folder is a Git repository is interesting, though it doesn't explain why its status is being refreshed at that moment.

Do one more thing: edit that orange breakpoint to say `console.log('repo:', repo.getPath()); console.trace(); false` and then reproduce the freeze. _Focus only on the log lines that appear to happen during the freeze_  (or right after it unfreezes). After each `repo:` line, there'll now be a stack trace. Examining them should shed light on what the mystery function is; the lines above `refreshStatus` should help determine where (and why) it's being called.

**Comment by Xon77** [comment]
The fact that my entire home folder is a Git repository should be an error when I try make pull request, but I choose the wrong directory for a specific project. And it stays there...

Anyway, below the stack trace. Hope it helps : 

repo: /Users/xon/.git/
VM4698:1 console.trace
eval @ VM4698:1
refreshStatus @ /Applications/Pulsar.app/Contents/Resources/app.asar/src/git-repository.js:563
onWindowFocus @ /Applications/Pulsar.app/Contents/Resources/app.asar/src/git-repository.js:98
VM4699:1 

repo: /Users/xon/atom-hydra/.git/
VM4699:1 console.trace
eval @ VM4699:1
refreshStatus @ /Applications/Pulsar.app/Contents/Resources/app.asar/src/git-repository.js:563
onWindowFocus @ /Applications/Pulsar.app/Contents/Resources/app.asar/src/git-repository.js:98

**Comment by savetheclocktower** [comment]
If these lines are truly happening during the freeze, then that's bizarre. It suggests that focusing the window is what's causing the freeze, but I would imagine that you're not switching between windows in the middle of this (right)?

> The fact that my entire home folder is a Git repository should be an error when I try make pull request, but I choose the wrong directory for a specific project. And it stays there...

**Comment by Xon77** [comment]
Well, during the freeze, I was not focusing the window.

But, GOOD NEWS : 

By removing the Git repository in my home folder, I can no longer reproduce the freeze right now.
I still have to test thoroughly to be sure that it does not happen with time again.

I am so happy, but also very confused because it seems to be due to "my bad handling of git".
I really owe you a beer in this virtual world.
Thank you so much for your patience.
It is so satisfying not to freeze!

I have not tested yet with the original version of Hydra, because you changed something with ndjson, which also prevented error messages of formatting. But, with the new version, it seems to play smoothly. I will give feedback about that asap.

I also noticed a lot of errors or deprecations when opening Pulsar, coming back to original post, which might also the cause of excess CPU at the beginning that I post below. But maybe it is known and will be solved in another release. And one thing at a time. A little bit too excited that you solved my issue!

Anyway, my freeze seems resolved for me until now. I cross the fingers.



electron/js2c/renderer_init.js:13 (electron) The remote module is deprecated. Use https://github.com/electron/remote instead.
log @ electron/js2c/renderer_init.js:13

electron/js2c/renderer_init.js:33 The vm module of Node.js is deprecated in the renderer process and will be removed.
o._load @ electron/js2c/renderer_init.js:33

/Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/scheduler/cjs/scheduler.development.js:348 
[Deprecation] SharedArrayBuffer will require cross-origin isolation as of M91, around May 2021. See https://developer.chrome.com/blog/enabling-shared-array-buffer/ for more details.
(anonymous) @ /Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/scheduler/cjs/scheduler.development.js:348

/Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/react-dom/cjs/react-dom.development.js:27773 
Download the React DevTools for a better development experience: https://fb.me/react-devtoolsYou might need to use a local HTTP server (instead of file://): https://fb.me/react-devtools-faq

/Applications/Pulsar.app/Contents/Resources/app.asar/src/styles-element.js:107 
[Deprecation] ::-webkit-details-marker pseudo element selector is deprecated. Please use ::marker instead. See https://chromestatus.com/feature/6730096436051968 for more details.
styleElementAdded @ /Applications/Pulsar.app/Contents/Resources/app.asar/src/styles-element.js:107

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/underscore/underscore-node-f.cjs.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/underscore/underscore-node.cjs.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/focus-trap/dist/focus-trap.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/tabbable/dist/index.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/dugite/build/lib/index.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/dugite/build/lib/git-process.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/dugite/build/lib/git-environment.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/dugite/build/lib/errors.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/atom-select-list/lib/select-list-view.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/uri-js/dist/es5/uri.all.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/performance-now/lib/performance-now.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/spelling-manager/lib/index.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/spelling-manager/lib/BufferSpellingChecker.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/spelling-manager/lib/regex-helper.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/spelling-manager/lib/SpellingManager.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/spelling-manager/lib/TokenCheckStatus.js.map: Unexpected end of JSON input

DevTools failed to load SourceMap: Could not parse content for file:///Applications/Pulsar.app/Contents/Resources/app.asar/node_modules/spelling-manager/lib/TokenSpellingManager.js.map: Unexpected end of JSON input

**Comment by savetheclocktower** [comment]
Glad to hear it!

All those warnings in the console are expected.

> I will give feedback about the original version asap.

Great, just make sure you back up your changes so you can restore your version of `atom-hydra` if need be.

**Comment by Xon77** [comment]
Well, apparently, the freeze does no longer happen with the original version of Hydra.
I Will continue to test to be sure and will report if I see anything.
It seems the git repository in my home folder generated all this mess (freeze and message error in json formatting).
Sorry and thanks again!

**Comment by savetheclocktower** [comment]
> I was never using Pulsar as a project editor; only a text editor. Mostly I had files open in my Downloads folder, but I may have had some files open that were nested inside a folder with a `.git`. Could that have been leading it to try refreshing the git status of every subdirectory it could find?

@Symbitic, I'm a bit mortified that we've probably been spamming your inbox; I should've taken our discussion to another issue long ago. Still, what we discovered with @Xon77 lends _some_ credence to this theory.

If your Downloads folder is the project directory (as I think it would be if you opened a single file in `~/Downloads` for editing), then I'd expect Pulsar's Git-repo logic to trigger only if `~/Downloads` itself is a Git repo — or perhaps if `~` is a Git repo, as in @Xon77’s case. (It's not always accidental; I know some people treat their home folder as a git repo for synchronization purposes.)

If `~/Downloads` merely _contains_ folders that might themselves be Git repositories, but does not itself descend from a Git repo, then I'm pretty sure Pulsar wouldn't just crawl the entire subtree looking for Git repos in speculative fashion.

You say it happens even when no files are open — which I imagine means with an empty window pointed at a certain folder. In that case, the only thing that matters is the location of the folder. Hence one way to test this would be to create a new folder somewhere on disk that is guaranteed to have no Git repos above or below it — somewhere in `/var`, perhaps — and see if you have the same performance problem when that's the only project open in Pulsar.

And whatever the outcome of that experiment is, I'm still interested in a startup profile if you want to attempt one.

Either way, though, the moral of this story is that Pulsar probably needs a faster way to get updated Git repo status, or else an asynchronous way that works better for repos that are potentially very large.

**Comment by 0ui0** [comment]
I have same issue

<img width="833" height="444" alt="Image" src="https://github.com/user-attachments/assets/1348a122-efe6-4c54-a68b-b2472970d17c" />

**Comment by savetheclocktower** [comment]
> I have same issue

What version of Pulsar? Is this new after an update, or has it been happening for a while?

**Comment by 0ui0** [comment]
> > I have same issue
> 
> What version of Pulsar? Is this new after an update, or has it been happening for a while?

No, this issue seems to have persisted for quite a while. It was already happening in versions from late last year. My macOS version is 15.7.3 (24G419). I encountered this even before the Electron update. I thought updating Electron would fix it, but unfortunately, the problem remains.
Usually, what happens is that I open my project, read some files, and then leave it running in the background (or sometimes I just launch it without even opening a project, and then leave it). I go do other things, and then a few hours later I find my MacBook extremely hot. I don't know how long it had been going on; I didn't notice until I realized the computer was acting up. I opened Activity Monitor and found the situation shown in the screenshot.
Because of this issue, Pulsar is almost unusable for me. Also, sometimes even after I've closed the foreground UI of Pulsar by clicking the red 'X', the app doesn't fully quit on macOS. At that point, I assume it's in some sort of dormant state. Even in this scenario, the issue still occurs.

**Comment by savetheclocktower** [comment]
I'm sorry this is happening. I would suggest two things:

1. Open your typical project [in safe mode](https://docs.pulsar-edit.dev/troubleshooting-pulsar/using-safe-mode/) and leave it in the background for a while. See if you can reproduce this issue without any community packages running.
2. If the issue persists after step 1, try creating a temporary directory with only a few items in it, and opening that as the only project. Leave that running for a few hours and see if any process starts using 100% of the CPU.

For other people, one common cause seems to be complexity of the filesystem underneath the project. For instance: one user had their whole home folder as the project root, and the library we use for filesystem events got confused with symlinks and kept traversing directories in cyclical fashion. If this is the cause in your case, then step 2 might help confirm it. We plan to fix this by switching to a new file-watching library in the near future.

I know it's not a lot of fun to run vague tests like these, but it does help us isolate a root cause. On the other hand, if you can reproduce the high CPU usage even in both the scenarios above, that's also useful information. The next steps would be a performance profile (to rule out the Chromium stack as the source of the high CPU usage) and attempting to attach a Node debugger (to troubleshoot whether something in the Node stack is getting confused).

**Comment by 0ui0** [comment]
> I'm sorry this is happening. I would suggest two things:
> 
> 1. Open your typical project [in safe mode](https://docs.pulsar-edit.dev/troubleshooting-pulsar/using-safe-mode/) and leave it in the background for a while. See if you can reproduce this issue without any community packages running.
> 2. If the issue persists after step 1, try creating a temporary directory with only a few items in it, and opening that as the only project. Leave that running for a few hours and see if any process starts using 100% of the CPU.
> 
> For other people, one common cause seems to be complexity of the filesystem underneath the project. For instance: one user had their whole home folder as the project root, and the library we use for filesystem events got confused with symlinks and kept traversing directories in cyclical fashion. If this is the cause in your case, then step 2 might help confirm it. We plan to fix this by switching to a new file-watching library in the near future.
> 
> I know it's not a lot of fun to run vague tests like these, but it does help us isolate a root cause. On the other hand, if you can reproduce the high CPU usage even in both the scenarios above, that's also useful information. The next steps would be a performance profile (to rule out the Chromium stack as the source of the high CPU usage) and attempting to attach a Node debugger (to troubleshoot whether something in the Node stack is getting confused).

It's possible. My project's directory structure is indeed quite complex, with a mix of node_modules and Cordova files. In some cases, the issue is hard to spot, which is why I didn't notice it after the update. But then it suddenly popped up again today.

**Comment by confused-Techie** [comment]
Just to add to this, as someone that hasn't seen this problem, that even if I try to recreate having a Pulsar project open for an extreme amount of time, I still can't recreate the CPU issues.

I've had a Pulsar project open for just over a week now on my work PC and according to Windows Task Manager it's using 0.1% CPU.

But this is on Windows 11, and opening a somewhat simple NodeJS [project](https://github.com/confused-Techie/org-map).

So just wanting to stress the inconsistencies with the behavior we are seeing here.

I wonder @savetheclocktower if there's any kind of additional logging we could attempt to implement and then collect to determine what's going on here?

**Comment by savetheclocktower** [comment]
I don't have great ideas for logging. If it's truly stuck in an infinite loop, adding logging would help us figure out where it's stuck… but here we don't have much of a hint about _where_ to add the logging. Even if my theory is correct that it's about file-watching — in that scenario I'd expect that the hot loop would be somewhere in the `nsfw` library, whether in the native code or in JS.

In my experience, it's easier to troubleshoot runaway CPU usage in an isolated Node process than it is in an Electron renderer process; this is one reason why #1572 altogether moves file-watching to a worker process.

@0ui0, if you're interested in helping test #1572, you can download one of the macOS binaries [produced by this job](https://github.com/pulsar-edit/pulsar/actions/runs/26203634471?pr=1572) and temporarily use that version of Pulsar. If, in that version, the _same renderer process_ eventually starts using lots of CPU, then this probably has nothing to do with file-watching. But if the high CPU shifts to a worker process, that'd be a strong indicator that we're on the right track. You'd then be able to switch the `core.fileSystemWatcher` setting to `@parcel/watcher` and determine if that fixes your issue.

**Comment by mangelmauriz** [comment]
Hi, 

I'm also affected by this issue on an Apple Silicon Mac. I've tried launching Pulsar in safe mode and also setting the file watcher to `@parcel/watcher` in [#1572](https://github.com/pulsar-edit/pulsar/pull/1572) as suggested by @savetheclocktower's to no avail.

What puzzles me is that I've tried @asiloisad's fork [lumine](https://github.com/lumine-code/lumine) and the issue isn't present here. From what I can tell, Lumine uses Node.js v24. Could that be what's making the difference?

**Comment by asiloisad** [comment]
I don’t think I know which change might help you sort this out, as I use Windows and have never come across this issue before.

**Comment by savetheclocktower** [comment]
> I'm also affected by this issue on an Apple Silicon Mac. I've tried launching Pulsar in safe mode and also setting the file watcher to `@parcel/watcher` in [#1572](https://github.com/pulsar-edit/pulsar/pull/1572) as suggested by [@savetheclocktower](https://github.com/savetheclocktower)'s to no avail.
> 
> What puzzles me is that I've tried [@asiloisad](https://github.com/asiloisad)'s fork [lumine](https://github.com/lumine-code/lumine) and the issue isn't present here. From what I can tell, Lumine uses Node.js v24. Could that be what's making the difference?

Possibly — and part of me hopes so, because it suggests that this will magically be fixed eventually when we upgrade Electron to a recent enough version. But I'm also on an Apple Silicon Mac and this isn't happening to me, so I hope we can narrow it down a bit and fix this more quickly so you don't have to wait around for that eventual upgrade.

@mangelmauriz, does this happen on any project, or only with certain ones? Since you tried safe mode, I'll rule out that it has anything to do with community packages, but one thing that we know to be problematic is certain directory structures.

(The theory that switching to `@parcel/watcher` might fix it came from my experience troubleshooting with a user who was treating their entire home folder as a project folder. That folder had lots of nested directories, some of which were symlinked to other locations, and some of which may even have been symlinked to ancestors of themselves such that a naïve file-watcher could keep descending indefinitely and get confused. I doubt that's happening in your case, but if you can reproduce the high CPU even when the only project open is a single folder with one or two files in it, then that tells me the issue is probably not related to file-watching.)

I will assume that, if you're savvy enough to know about Lumine and to have tried it out, you're also savvy enough to have downloaded the specific build for which `@parcel/watcher` is a file-watching option. No matter which file-watcher you pick on that build, the other major thing it does is move file-watching into its own process; so if the same process was using excess CPU as before (rather than one explicitly marked as a file-watching process in Activity Monitor) then that's further evidence that this has nothing to do with file-watching.

The remaining theories that come to mind:

1. Git/GitHub integration.
2. Module caching issues.

### Git/GitHub integration

I think it's happened in the past that certain users have run into high CPU usage because of tasks related to Git communication. Anything that uses our `git-utils` library and the `libgit2` bindings is a bridge into a C++ library and could possibly result in high CPU usage whose origin not immediately obvious upon inspecting a performance profile.

You can do two things to rule out this theory: first, temporarily disable the built-in `github` package and relaunch Pulsar. Second: as I directed above, test on a small project with only a few files, but which is not the root of a Git repo, nor within a Git repo at all. (And as happened to the user who reported this issue… it would be good to double-check that your entire home folder is not under version control. Certainly Pulsar should work well even if you've made that choice, but if that's not the issue then it's worth ruling out for good.)

###  Module caching issues

The module cache is one of the least understood parts of Pulsar. I have long suspected that it is no longer truly necessary, but have not had the bandwidth to run experiments. In the past I've run into symptoms like these when switching from running an official release to running a one-off unsigned CI build; performance profiling suggested that _practically everything_ was an order of magnitude slower than it should've been, and I seem to remember it fixed itself when I cleared the module cache.

Do this first: back up your entire `~/.pulsar` directory (or wherever else `ATOM_HOME` points). This will free you up to run some bold experiments and then revert them simply by restoring the entire contents of the original folder.

Once you have a backup in place, I'd first suggest you remove everything within the `compile-cache` folder — all the subdirectories and everything. Then relaunch (always testing against a minimal project as before) and see if the problem persists.

If so, I'd go further and actually delete `~/.pulsar` in its entirety and relaunch; that would restore you to the first-launch state, more or less. If that doesn't work, you can abandon this experiment and restore your original `~/.pulsar` from the backup.

If this helps at all, then it suggests the real win was not necessarily moving to Lumine, but rather moving to a new application with a new home-folder location (I assume!) and a from-scratch experience. That would further point to an issue with the compile cache.

### Other options

* If you still strike out, then the only other thing I can recommend is to capture a performance profile and send it to me (or attach it to this issue). Even if it doesn't reveal anything obvious, it could give me hints as to what's going on.

* I also would be curious to find out what happens if you select the 100%-CPU-using item in Activity Monitor and quit it (or force quit it if necessary). I expect that process to be named “Pulsar Helper (Renderer),” but that doesn't necessarily mean it's the main renderer process for a window — because I believe that a new Node process spawned from a renderer window inherits the parent process’s title unless it's explicitly changed to something else.

  So if you kill the process and your project window complains about a crash, then the main renderer process was the one using all the CPU, and that suggests a performance profile from the dev tools is likely to have useful information. But if you kill the process and you _don't_ see a crash dialog, then it was likely to be a process spawned by a built-in package for a certain task, or else a secret third thing whose identity will be fun to discover. Either way we can narrow it down from there. Let me know which of these happens!

Thanks for reporting this, @mangelmauriz!

**Comment by mangelmauriz** [comment]
I've checked, and it turned out I had a Git repository in my home directory 🤦. Removing it solved the problem, and now my CPU usage is back to normal.

After solving the problem, I initialized a Git repository in my home directory again to investigate it further. After disabling the `git-diff`, `github`, `language-git`, and `open-on-github` packages and restarting Pulsar, the high CPU usage still remains, which seems odd to me.

Also, I tried killing the process with the abnormal CPU usage, `Pulsar Helper`, as you suggested, and I got a **"The editor has crashed"** window.

If you'd like a performance profile or anything else, I'd be happy to help. Thanks for your thorough answer, @savetheclocktower.

**Comment by savetheclocktower** [comment]
Hey, that's a far happier ending than I had figured we'd get!

I think the Git issue _might_ end up being fixed once we update our `git-utils` library to use a newer version of `libgit2` — which I hope will happen in the next release. (@asiloisad, did you bring that fix into Lumine? That would explain it.)

**Comment by asiloisad** [comment]
Lumine has dropped the use of git-utils and now uses the system Git only. It may the point, but..

...I updated the vendors before dropping these dependencies:
- https://github.com/lumine-code/git-utils <- libgit2 updated to 1.9.4
- https://github.com/lumine-code/node-pathwatcher <- efsw updated to 1.6.3

You may want to pin it temporary and perform smoke tests.

**Comment by savetheclocktower** [comment]
Thanks.

@mangelmauriz, upon re-reading your comment, this stands out:

> After disabling the `git-diff`, `github`, `language-git`, and `open-on-github` packages and restarting Pulsar, the high CPU usage still remains, which seems odd to me.

To me, that strengthens the case that this is a `git-utils` issue one way or another. Even with those packages disabled, Pulsar has Git awareness baked into the core. (Example: `atom.project.getRepositories()` returns any number of `GitRepository` instances, each of which wraps a `git-utils` `Repository` instance. Any package can use this API; `status-bar` uses it to display your repository status.)

We'll find out soon enough!
